src/core/hle/D3D8/XbPixelShader.cpp

```diff
@@ -638,6 +642,297 @@ constexpr int PSH_XBOX_CONSTANT_FRONTFACE_FACTOR = PSH_XBOX_CONSTANT_LUM + 4; //
 // This concludes the set of constants that need to be set on host :
 constexpr int PSH_XBOX_CONSTANT_MAX = PSH_XBOX_CONSTANT_FRONTFACE_FACTOR + 1; // = 28
 
+std::string GetFixedFunctionShaderTemplate() {
+	static bool loaded = false;
+	static std::string hlslString;
+
+	// TODO does this need to be thread safe?
+	if (!loaded) {
+		loaded = true;
+
+		// Determine the filename and directory for the fixed function shader
+		auto hlslDir = std::filesystem::path(szFilePath_CxbxReloaded_Exe)
+			.parent_path()
+			.append("hlsl");
+
+		auto sourceFile = hlslDir.append("FixedFunctionPixelShader.hlsl").string();
+
+		// Load the shader into a string
+		std::ifstream hlslStream(sourceFile);
+		std::stringstream hlsl;
+		hlsl << hlslStream.rdbuf();
+
+		hlslString = hlsl.str();
+	}
+
+	return hlslString;
+}
+
+std::string_view GetD3DTOPString(int d3dtop) {
+	const std::string opToString[] = {
```

Comment:
```suggestion
	static constexpr std::string_view opToString[] = {
```

Those are constant strings so no need to exercise heap allocations. `string_view` will most likely make them compiled in .rdata as-is.